crates/runtime/src/externref.rs

```diff
@@ -0,0 +1,445 @@
+            });
+
+        let alloc_align = std::cmp::max(value_align, extern_data_align);
+        let alloc_size = value_and_padding_size
+            .checked_add(extern_data_size)
+            .unwrap_or_else(|| Self::alloc_failure());
+
+        unsafe {
+            debug_assert!(Layout::from_size_align(alloc_size, alloc_align).is_ok());
+            let layout = Layout::from_size_align_unchecked(alloc_size, alloc_align);
+
+            let alloc_ptr = std::alloc::alloc(layout);
+            let alloc_ptr = NonNull::new(alloc_ptr).unwrap_or_else(|| {
+                Self::alloc_failure();
+            });
+
+            let value_ptr = alloc_ptr.cast::<T>();
+            ptr::write(value_ptr.as_ptr(), make_value());
+
+            let value_ref: &T = value_ptr.as_ref();
+            let value_ref: &dyn Any = value_ref as _;
+            let value_ptr: *const dyn Any = value_ref as _;
+            let value_ptr: *mut dyn Any = value_ptr as _;
+            let value_ptr = NonNull::new_unchecked(value_ptr);
+
+            let extern_data_ptr =
+                alloc_ptr.cast::<u8>().as_ptr().add(value_and_padding_size) as *mut VMExternData;
+            ptr::write(
+                extern_data_ptr,
+                VMExternData {
+                    ref_count: UnsafeCell::new(1),
+                    value_ptr,
+                },
+            );
+
+            VMExternRef(NonNull::new_unchecked(extern_data_ptr))
+        }
+    }
+
+    /// Turn this `VMExternRef` into a raw, untyped pointer.
+    ///
+    /// This forgets `self` and does *not* decrement the reference count on the
+    /// pointed-to data.
+    ///
+    /// This `VMExternRef` may be recovered with `VMExternRef::from_raw`.
+    pub fn into_raw(self) -> *mut u8 {
+        let ptr = self.0.cast::<u8>().as_ptr();
+        mem::forget(self);
+        ptr
+    }
+
+    /// Create a `VMExternRef` from a pointer returned from a previous call to
+    /// `VMExternRef::into_raw`.
+    ///
+    /// # Safety
+    ///
+    /// Wildly unsafe to use with anything other than the result of a previous
+    /// `into_raw` call!
+    ///
+    /// This method does *not* increment the reference count on the pointed-to
+    /// data, so `from_raw` must be called at most *once* on the result of a
+    /// previous `into_raw` call. (Ideally, every `into_raw` is later followed
+    /// by a `from_raw`, but it is technically memory safe to never call
+    /// `from_raw` after `into_raw`: it will leak the pointed-to value, which is
+    /// memory safe).
+    pub unsafe fn from_raw(ptr: *mut u8) -> Self {
+        debug_assert!(!ptr.is_null());
+        VMExternRef(NonNull::new_unchecked(ptr).cast())
+    }
+
+    #[inline(never)]
+    #[cold]
+    fn alloc_failure() -> ! {
+        panic!("VMExternRef allocation failure")
+    }
+
+    #[inline]
+    fn extern_data(&self) -> &VMExternData {
+        unsafe { self.0.as_ref() }
+    }
+}
+
+impl PartialEq for VMExternRef {
```

Comment:
This and traits below seem dangerous since I'd naively expect it to defer to the underlying type (which can't be done here b/c it's not generic). Could this be exposed as a free-function?